### ISSUE TITLE
API types contain __typename, queries do not

### ISSUE THREAD

**Issue by eettaa** [aws-amplify/amplify-codegen#445]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-codegen/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-codegen/blob/master/CONTRIBUTING.md#bugs).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

npm 

### If applicable, what version of Node.js are you using?

14.18.1

### Amplify CLI Version

7.6.21 (sorry, I'm blocked by https://github.com/aws-amplify/amplify-category-api/issues/560)

### What operating system are you using?

Ubuntu

### Amplify Codegen Command

codegen

### Describe the bug

API.ts as generated by codegen has typescript types that include a string literal __typename field, e.g.
```
export type GetSavingsAccountQuery = {
  getSavingsAccount?:  {
    __typename: "SavingsAccount",
    accountId: string
    ...
}
```
But the associated graphql query in queries.ts does NOT query the __typename field:
```
  query GetSavingsAccount($accountId: ID!) {
    getSavingsAccount(accountId: $accountId) {
      accountId
      (no __typename here!)
      ...
  }
}
```

This means that trying to access __typename on a query response will always typecheck but never succeed at runtime (because that field is never returned).

Some frontend graphql clients like ApolloClient automatically append __typename to all queries, which works around this problem.  however, if you were to use the literal query as generated, it would not match the type as generated.

Is there an option to have codegen add __typename to all the queries?

### Expected behavior

all queries in queries.ts also query the __typename field, or (less preferrable to me), API.ts does not export __typename as a field in all query responses.

### Reproduction steps

Take any trivial @model-annotated type in schema
Codegen it
Inspect API.ts and queries.ts, notice one has __typename and one does not.

### GraphQL schema(s)

I have instead included a graphqlconfig.yml:
```
projects:
  projectname:
    schemaPath: path/to/schmea
    includes:
      - path/to/codegen/**/*.ts
      - path/to/custom/**/*.ts
    excludes:
      - ./amplify/**
    extensions:
      amplify:
        codeGenTarget: typescript
        generatedFileName: path/to/codegen/API.ts
        docsFilePath: path/to/codegen
        region: us-west-2
        apiId: null
        maxDepth: 6
extensions:
  amplify:
    version: 3
```

### Log output

<details>

```
# Put your logs below this line


```

</details>

**Comment by hisham** [comment]
+1. Any update on this? Is this a regression? I'm pretty sure __typename was included before...

**Comment by ossterv** [comment]
+1 Same issue here. I used amplify-cli version 10.0.0. Only the first level of query or mutation has the __typename defined. Subtypes are having it as a string, not named type. Yes, it is regression compared to amplify-cli version 4.29.4, which we were using before. The workaround is to use (lots of) typescript's typecasting, which pollutes our code.

**Comment by eettaa** [comment]
Ping here again, please help!  An engineer on my time just spent many hours re-discovering this bug ;(

**Comment by eettaa** [comment]
@josefaidt @AnilMaktala @AaronZyLee Do any of you have a minute to look at this? It has stalled in pending triage for nearly seven months, and my additional pleas for attention in duplicate https://github.com/aws-amplify/amplify-category-api/issues/898 were acknowledged without action...
Thanks!!

**Comment by AaronZyLee** [comment]
@eettaa Sorry for the late response. The issue here looks like a missing gap between generated types vs GQL ops. From the comments above it was indicated as a regression but I am not able to generate ops with `__typename` by using older version CLI. It would be helpful if you could provide more info on whether this is a regression.

Meanwhile I will bring this out with team and review if there are any side effects by introducing the `__typename` in the generated code. If no breaking changes are foreseen, we will implement this ASAP.

If this is urgent issue for unblocking, a temporary solution is to add `__typename` manually in the generated code. 

Again sorry for the inconvenience and late response. Will get back to you once we make the decision.

**Comment by eettaa** [comment]
Hi @AaronZyLee , thanks for your response!
I do not know whether this is a regression or not- I was just reporting it as a bug (which I think it clearly is- the generated types should match the generated queries i.e. they should both include __typename or both not).
@hisham and @ossterv claimed it was a regression- do either of you have additional information?  I guess @ossterv believes it was working in amplify 4.29.4

**Comment by eettaa** [comment]
@AaronZyLee @josefaidt could we take another look at this please?  Regardless of whether or not it is a regression, I feel that it is an existing bug.

> Meanwhile I will bring this out with team and review if there are any side effects by introducing the __typename in the generated code. If no breaking changes are foreseen, we will implement this ASAP.

Did your internal team come back with a response here?

**Comment by AaronZyLee** [comment]
@eettaa Hi we decide to implement the feature and add this behind a feature flag in consideration of adding extra costs and payloads in the selection set. Will update with you once the PR is ready.

**Comment by sprkem** [comment]
Looks like this was reverted - if so can we reopen?

**Comment by dpilch** [comment]
Sorry, we had an issue with the release being incompatible with DataStore. At the latest we will be able to release this feature in conjunction with Amplify JS V6 coming later this year. We are looking into some other options to possibly release this sooner, but no promises yet.

**Comment by dpilch** [comment]
This has been released in v12.1.0 of the cli.
